update-source-git/README.md

```diff
@@ -0,0 +1,105 @@
+`packit source-git update-source-git`
+
+This research covers a command which synchronizes changes from dist-git back to
+a source-git repo.
+
+Areas to cover:
+
+- spec file synchronization
+
+- sources synchronization
+
+- `.distro/source-git.yaml` update
+
+- code synchronization
+
+Let's break these down.
+
+## Spec file sync
+
+This seems trivial, we can copy the file from dist-git back to source-git. But
+before we do that, we should capture the version field and check if it differs
+from the source-git spec.
+
+## Sources synchronization
+
+Not much to do here, except that we should detect a removal of a file and also
+remove it in source-git's `.distro` folder.
+
+## `.distro/source-git.yaml` update
+
+Unless the version changes, there should be nothing to update. Actually this
+file should always be changed first in the source-git and there should be no
+changes happening here when getting new content from dist-git.
+
+## Code synchronization
+
+There are multiple cases in this bucket, let's talk about them individually.
+
+Cases 2) and 3) require a rebase in source-git while 1) can be fast-forwarded.
+
+At first, we should be able to detect that such a thing happens. There are
+multiple ways how to do that - one particularly interesting is to inspect
+commits and check what files have changed: if sources file or patch files
+changed, we can easily tell source code changed: cases 2) and 3).
+
+We don't intend to support the "rebase" use case, people should still do work
+in their source-git repos and not dist-git. Although we intend to support
+proven-packager workflow: being able to synchronize simple spec file changes
+from dist-git to source-git.
+
+One problem with detecting changes is that we do not track which source-git
+commit matches a dist-git commit and vice versa. We should start with requiring
+a git-range which should be synchronized and over time figure out if this can
+be automated. Alternatively we can start putting metadata into our dist-git
+commits to tell which source-git commit they are matching.
+
+### New patch was added on top
+
+New patch was added in a dist-git repo on the of the code change history - we
+apply it on top of the source-git repo in a fast-forward way.
+
+### Code history changed
+
+1. A patch was removed
+2. A patch was altered in the commit history
+3. A patch was added within the commit history
+
+### Version changed
+
+The package was rebased to a different version which resulted into several
+patches being dropped.
+
+## CLI proposal
+
+```
+Usage: packit source-git update-source-git [OPTIONS] <REVISION RANGE> SOURCE_GIT DIST_GIT
```

Comment:
I think it would be more natural if the order of the arguments would match the direction in which data flows. This command will take data from dist-git and put it to source-git, so the order should be:

```suggestion
Usage: packit source-git update-source-git [OPTIONS] REVISION RANGE DIST_GIT SOURCE_GIT
```

Reply:
Logically that makes sense to me and I'm going to make that change (swap). My point originally was to be consistent in terms of the order of the arguments with the other update command. The reason was that I dislike in e.g. git that command arguments are very inconsistent between multiple commands.

One more change I'm going to do will be to move the `<REVISION RANGE>` to the end for two reasons:
1. it makes sense to run the command multiple times for different refs - the repo locations don't move, refs do
2. if we ever come up with a default for the range (e.g. default to HEAD), we can make it optional